### ISSUE TITLE
Demo using warp

### ISSUE THREAD

**Issue by joshtriplett** [stjepang/smol#91]
I managed to run `warp` in `smol`, quite easily:

```rust
use warp::Filter;

fn main() {
    for _ in 0..num_cpus::get() {
        std::thread::spawn(|| smol::run(futures::future::pending::<()>()));
    }

    smol::block_on(async {
        let hello = warp::path!("hello" / String)
            .map(|name| format!("Hello, {}!", name));

        warp::serve(hello)
            .run(([127, 0, 0, 1], 8080))
            .await;
    })
}
```

Please consider adding this as a demo.

**Comment by Stupremee** [comment]
Why do you spawn some threads at the beginning that will just execute the pending future?

**Comment by tekjar** [comment]
@Stupremee That's how multi threaded runtime is setup in smol. I haven't gone through the internals yet but smol has work stealing which distributes tasks to these threads

**Comment by szagi3891** [comment]
This is amazing. I switched a few lines and my project using wrap and it just started :D

**Comment by csarn** [comment]
For me it panics at 
    there is no reactor running, must be called from the context of Tokio runtime

it seems to come from hyper::server::tcp::AddrIncoming::from_std calling something from tokio.
How would I address that?

**Comment by leo-lb** [comment]
@csarn See here: https://github.com/stjepang/smol/blob/master/examples/hyper-server.rs

hyper uses tokio internally by default, this uses an interface that hyper offers for runtime-agnostic operation, to use hyper (which warp uses) with smol.

**Comment by leo-lb** [comment]
@joshtriplett I think you example still uses Tokio under the hood. So you are running Tokio *AND* smol, which is not great. You would need to intialize the HTTP server with a custom acceptor that uses smol like explained in my previous comment.

**Comment by szagi3891** [comment]
@csarn 
does your dependence to "smol" look like this?
smol = { version = "0.1", features = ["tokio02"] }

**Comment by csarn** [comment]
@szagi3891 ha thanks, I didn't see that part of the README :) I looked at the github repo a while back when the README was still empty, and didn't think of looking again... 
Now it works.

**Comment by szagi3891** [comment]
With smol you can run code that uses async-std and tokio libraries. For me, this is the most appealing feature.

Here is more explained by the author:
https://stjepang.github.io/2020/04/03/why-im-building-a-new-async-runtime.html

**Comment by leo-lb** [comment]
@csarn @szagi3891

If you run a library with an explicit dependency on Tokio with smol, there's clearly no advantage to use smol. The tokio compatibility layer is provided to help migration but it's clearly expected for libraries to become runtime-agnostic. Running two runtimes in the same program is undesirable, unless used temporarily until migration happens, it's not great.

**Comment by zicklag** [comment]
I opened an issue in Warp to ask about this: https://github.com/seanmonstar/warp/issues/601#issue-635791319.

**Comment by leo-lb** [comment]
@zicklag seeing that they closed your issue I encourage you fork and modify this file:

https://github.com/seanmonstar/warp/blob/master/src/server.rs

You can inspire yourself from the hyper server example here:

https://github.com/stjepang/smol/blob/master/examples/hyper-server.rs

**Comment by zicklag** [comment]
Oh, great. Thanks for the pointer. I was wondering about that, but I needed a hint in the right direction. If anybody else wants to try it first, I'm not sure when I'll get to it. :smiley:

**Comment by kvsari** [comment]
I might give this a shot. It looks like adding an extra method [like this one](https://github.com/seanmonstar/warp/blob/master/src/server.rs#L337).

I imagine some extra parameters to be sent such as the spawner to be used as is happening [here](https://github.com/smol-rs/smol/blob/master/examples/hyper-server.rs#L43). This spawner is way to use another futures runtime.

Please correct me if I'm wrong.

**Comment by kvsari** [comment]
I have a quick and dirty [fork](https://github.com/kvsari/warp/tree/spawner). I've had to resort to using a newtype, `NtExec`. I couldn't figure out how to make it a trait bound with `E: hyper::rt::Executor`. It'll do for now to prove whether it can work or not.

To use you'll have to do something like this;
```rust
pub struct WarpSpawn(futext::spawner::Spawner);

impl warp::hyper::rt::Executor<Pin<Box<dyn Future<Output = ()> + Send + 'static>>> for WarpSpawn
{
    fn execute(&self, fut: Pin<Box<dyn Future<Output = ()> + Send + 'static>>) {
        self.0.spawn(async { drop(fut.await) }).detach();
    }
}
```
The `futext` is a small [lib](https://git.sr.ht/~kvsari/futext) I'm playing with. It's all `smol` underneath.

Then later inside main...
```rust
let warp_spawn = WarpSpawn{ 0: rt.spawner() };
let warp_exec = warp::NtExec{ 0: Arc::new(warp_spawn) };
```

Again, this all compiles but it doesn't work for me. I'm using websockets in my project and lo and behold, warp [hardcodes tokio](https://github.com/seanmonstar/warp/blob/master/src/filters/ws.rs#L151). I'd have to figure out a way to thread the passed in executor into the `ws` module. I don't have the time for this right now unfortunately. 

Someone else can have a look and see if at least the pure HTTP stuff works.
